Handle entities that have been removed on the server

seadowg opened the pull request:
Closes #6011


#### Why is this the best possible solution? Were any other approaches considered?

The big change here is that our locally stored entities (of type `Entity`) can now be "offline" (they only exist locally) or "online" (they exist locally and on the server). This concept refers to the entities ID rather than its version: we might have made an update locally that's not on the server yet to an entity we know is "online". Entities become "online" once they've been seen in a form's entity list. As part of this change, I've updated the language we use in the code to reflect it better.

As noted in the issue, there isn't a way that we could handle one of the cases (a form deleted on the server that has never been "seen" by the client in an entity list) without adding new communication with the server. This is something that's being actively discussed with the Central team. 

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

Not much has changed other than entities so testing can probably be restricted to that. 

One thing to note is that entities will likely be cleared when upgrading from one version of Collect to another with the current implementation. As discussed [here](https://github.com/getodk/collect/pull/6132#discussion_r15983850970), this is just a temporary caveat while the features are experimental so allow us to iterate faster.

#### Before submitting this PR, please make sure you have:
- [x] added or modified tests for any new or changed behavior
- [x] run `./gradlew connectedAndroidTest` (or `./gradlew testLab`) and confirmed all checks still pass
- [x] added a comment above any new strings describing it for translators
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

seadowg committed:
Update naming for local entities

"local" entities are ones stored locally and can include "offline" and "online" ones.

seadowg committed:
Add test for updating entities with missing online one

seadowg committed:
Remove online entities that no longer appear in list

seadowg committed:
Make sure online entities updated offline remain online

seadowg committed:
Show pill on offline entities

seadowg committed:
Move custom theme attribute

seadowg committed:
Style offline pill

seadowg committed:
Emphasise entity label

seadowg committed:
Fix entity delete in cases where there are no updates

seadowg committed:
Simplify local entity logic

seadowg committed:
Fix error pill

seadowg committed:
Extract entity parse logic

seadowg committed:
Make entity naming consistent

seadowg committed:
Update test naming guidelines

seadowg committed:
Update tests to reflect that entities have unique IDs across lists

seadowg committed:
Clear entities if they fail to parse

seadowg commented on collect_app/src/test/java/org/odk/collect/android/entities/JsonFileEntitiesRepositoryTest.kt:
```diff
@@ -23,8 +26,22 @@ class JsonFileEntitiesRepositoryTest : EntitiesRepositoryTest() {
+    }
+
+    @Test
+    fun `deletes existing backing file if it can't be parsed by current code`() {
```
This prevents us running into crashes when upgrading versions that use `JsonFileEntitiesRepository`. Clearing out entities on upgrade is obviously not ideal, but we don't want to go down the road of writing JSON migrations or storing the in a DB (quite yet). We'll note this as a caveat to users when the beta gets released.

grzesiek2010 replied:
Actually, you assert that the repository no longer contains the entity but not if the file exists or not so it's not consistent with what the title says.

seadowg commented on docs/CODE-GUIDELINES.md:
````diff
@@ -28,6 +28,29 @@ assertThat(ClassToTest.methodToTest("input"), equalTo("expected"));
 assertThat(ClassToTest.methodReturnsNull(), equalTo(null));
 ```
 
+### Backtick test naming
````
This section is definitely all up for debate. I'd just noticed that we hadn't really found a consistent style, so wanted to propose one for discussion!

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/entities/JsonFileEntitiesRepository.kt:
```diff
@@ -11,29 +11,36 @@ class JsonFileEntitiesRepository(directory: File) : EntitiesRepository {
 
             if (existing != null) {
+                val offline = if (existing.offline) {
+                    entity.offline
```
It could be just true.

seadowg replied:
This totally fooled me as well looking at it again. If `existing` is offline then we want to use the new entitie's (`entity`) offline value, but if not we ignore it (we never go from "online" to "offline"). You can demo this by changing it to `true`: a test will fail. I think the change you suggest in https://github.com/getodk/collect/pull/6132#discussion_r1604442916 could hopefully make this all a little easier to read.

grzesiek2010 commented on entities/src/main/java/org/odk/collect/entities/InMemEntitiesRepository.kt:
```diff
@@ -2,40 +2,51 @@ package org.odk.collect.entities
 
             if (existing != null) {
+                val offline = if (existing.offline) {
+                    entity.offline
```
It could be just true.

grzesiek2010 commented on collect_app/src/test/java/org/odk/collect/android/entities/EntitiesRepositoryTest.kt:
```diff
@@ -57,40 +57,49 @@ abstract class EntitiesRepositoryTest {
-        val whiskys = repository.getEntities("whiskys")
-        assertThat(whiskys, contains(whisky))
+    @Test
+    fun `#save does not update offline on existing entity when it is true`() {
```
I find this title difficult to understand. Could it be:
`#save does not update offline on existing entity when it is already online`?
Maybe it would be even better to replace this boolean flag with an enum `OFFLINE`, `ONLINE`. Thanks to that the test title could be like:
`#save does not change entity state to OFFLINE when it is already ONLINE`

The same with the test above.

grzesiek2010 commented on material/src/main/java/org/odk/collect/material/MaterialPill.kt:
```diff
@@ -35,6 +36,15 @@ open class MaterialPill(context: Context, attrs: AttributeSet?) :
+            text = getString(R.styleable.MaterialPill_text)
+
+            val iconId = getResourceId(R.styleable.MaterialPill_icon, -1)
+            if (iconId != -1) {
```
Why would this be -1?

seadowg replied:
If `R.styleable.MaterialPill_icon` isn't defined `getResourceId` will return `-1` (as that's passed in as the default. Annoyingly, there isn't a nullable version of the method as far as I've seen.

grzesiek2010 commented on collect_app/src/androidTest/java/org/odk/collect/android/support/pages/EntitiesPage.kt:
```diff
@@ -1,15 +1,13 @@
     }
 
-    fun clickOnDataset(datasetName: String): DatasetPage {
+    fun clickOnList(datasetName: String): DatasetPage {
```
The parameters should be renamed too.

grzesiek2010 commented on collect_app/src/androidTest/java/org/odk/collect/android/support/pages/EntitiesPage.kt:
```diff
@@ -1,15 +1,13 @@
-    fun clickOnDataset(datasetName: String): DatasetPage {
+    fun clickOnList(datasetName: String): DatasetPage {
         clickOnText(datasetName)
         return DatasetPage(datasetName)
```
There is also `DatasetPage` class that could be renamed.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/entities/JsonFileEntitiesRepository.kt:
```diff
@@ -11,29 +11,36 @@ class JsonFileEntitiesRepository(directory: File) : EntitiesRepository {
 
     private val entitiesFile = File(directory, "entities.json")
 
-    override fun getDatasets(): Set<String> {
```
in `readEntities()` and `JsonEntity.toEntity(dataset: String)` there are still occurrences of `dataset` instead of `list`.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/entities/JsonFileEntitiesRepository.kt:
```diff
@@ -77,27 +90,37 @@ class JsonFileEntitiesRepository(directory: File) : EntitiesRepository {
+                mutableMapOf()
+            }
+        } catch (e: Exception) {
+            entitiesFile.delete()
```
I think you could have a separate method like:
```java
private fun deleteFile() {
         entitiesFile.delete()
         entitiesFile.createNewFile()
    }
```
You don't have to call #createFile as the parent directory should still exist right? so you just need to create the file again.

grzesiek2010 commented on collect_app/src/test/java/org/odk/collect/android/entities/EntitiesRepositoryTest.kt:
```diff
@@ -57,40 +57,49 @@ abstract class EntitiesRepositoryTest {
         repository.save(wine)
 
-        val updatedWine = wine.copy(label = "Léoville Barton 2009")
+        val updatedWine = wine.copy(label = "Léoville Barton 2009", version = 1)
```
You don't have to set the version number using copy if it's the same. Just the label has changed.

grzesiek2010 commented on lists/src/main/java/org/odk/collect/lists/RecyclerViewUtils.kt:
```diff
@@ -14,4 +16,11 @@ object RecyclerViewUtils {
         return itemDecoration
     }
+
+    fun RecyclerView.ViewHolder.setItemViewLayoutParams() {
```
The goal here is to expand the view horizontally right? If so it would be good to have a name that reflects the intention, for example: `expandItemViewHorizontally` or `stretchItemViewHorizontally`

grzesiek2010 commented on collect_app/src/test/java/org/odk/collect/android/entities/EntitiesRepositoryTest.kt:
```diff
@@ -63,32 +63,13 @@ abstract class EntitiesRepositoryTest {
     }
 
-    @Test
-    fun `#save does not update existing entity with matching id but not list`() {
```
Why is this testpassing (would be if it still existed) if entities should have unique id? Do we have a test for that scenario?

seadowg replied:
I think when I initially added this test I was thinking that two entities in different lists could have the same ID. This is actually not something we want to support though (entities should have unique IDs across lists), so I removed the test so as not to imply the opposite. I'll add a test to specify what happens if you pass `save` an entity with an existing ID in a different list.

seadowg committed:
Update test for JSON format changes

seadowg committed:
Remove other references to dataset

seadowg committed:
Simpplify file recreation

seadowg committed:
Remove unneeded param

seadowg committed:
Make method intention clearer

seadowg committed:
Specify behaviour when updating entities with same ID in different lists

seadowg committed:
Convert boolean to enum

grzesiek2010 commented on collect_app/src/androidTest/java/org/odk/collect/android/support/pages/EntityListPage.kt:
```diff
@@ -0,0 +1,15 @@
+package org.odk.collect.android.support.pages
+
+class EntityListPage(private val datasetName: String) : Page<EntityListPage>() {
```
The parameter should be `listName` instead of `datasetName` I think.

grzesiek2010 commented on collect_app/src/test/java/org/odk/collect/android/entities/JsonFileEntitiesRepositoryTest.kt:
```diff
@@ -23,8 +26,25 @@ class JsonFileEntitiesRepositoryTest : EntitiesRepositoryTest() {
+        val backingFile = filesInDir[0]
+        backingFile.writeText("blah")
+
+        assertThat(repository.getEntities("blah").size, equalTo(0))
```
Her it should be `stuff` not `blah` right?

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/entities/JsonFileEntitiesRepository.kt:
```diff
@@ -81,20 +93,26 @@ class JsonFileEntitiesRepository(directory: File) : EntitiesRepository {
+            }
+        } catch (e: Exception) {
+            entitiesFile.delete()
+            entitiesFile.createNewFile()
```
If I comment those two lines out this test still passes: https://github.com/getodk/collect/pull/6132/files#diff-f7e8798917d3ccfe8e394d1d714eb9535648a9c63a42c41b5ec30d32006cbd2cR35

seadowg replied:
Ah good spot! We don't actually need to delete the file itself as we just treat the file as empty and then override next time we mutate (as we always do anyway). I can remove the `delete`/`createNewFile` call.

seadowg committed:
Rename parameter

seadowg committed:
Fix test assertion

seadowg committed:
Remove unneeded file deletion

We write the whole file again when mutating, so treating the
file as empty is good enough.

srujner commented:
**Tested with success!**

Verified on a device with Android 13

Verified cases:
- Acceptance Criteria from #6011 has been met;
- Upgrading Collect;
- Ensuring the app correctly identifies entities that have been removed on the server.
- Testing the synchronization process to confirm that removed entities are not re-downloaded.
- Validating that any cached data for removed entities is cleared appropriately.
- Screen rotation, minimize app, using backward button,

WKobus commented:
**Tested with success!**

Verified on a device with Android 14

dbemke commented:
**Tested with success!**

Verified on a device with Android 10
